velox/exec/ArrowStream.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/ArrowStream.h"
+
+namespace facebook::velox::exec {
+
+ArrowStream::ArrowStream(
+    int32_t operatorId,
+    DriverCtx* driverCtx,
+    const std::shared_ptr<const core::ArrowStreamNode>& arrowStream)
+    : SourceOperator(
+          driverCtx,
+          arrowStream->outputType(),
+          operatorId,
+          arrowStream->id(),
+          "Arrow Stream") {
+  arrowStream_ = arrowStream->arrowStream();
+}
+
+ArrowStream::~ArrowStream() {
+  if (!isFinished()) {
+    close();
+  }
+}
+
+RowVectorPtr ArrowStream::getOutput() {
+  struct ArrowArray arrowArray;
+  if (arrowStream_->get_next(arrowStream_.get(), &arrowArray)) {
+    VELOX_FAIL(
+        "Failed to call get_next on ArrowStream: " + std::string(getError()))
+  }
+  if (arrowArray.release == NULL) {
+    // End of Stream.
+    finished_ = true;
+    return nullptr;
+  }
+  struct ArrowSchema arrowSchema;
+  // Get arrow schema.
+  if (arrowStream_->get_schema(arrowStream_.get(), &arrowSchema)) {
+    VELOX_FAIL(
+        "Failed to call get_schema on ArrowStream: " + std::string(getError()))
+  }
+
+  // Convert Arrow Array into RowVector and return.
+  return std::dynamic_pointer_cast<RowVector>(
+      importFromArrowAsOwner(arrowSchema, arrowArray, pool()));
+}
+
+void ArrowStream::close() {
+  arrowStream_->release(arrowStream_.get());
+  SourceOperator::close();
+}
+
+bool ArrowStream::isFinished() {
+  return finished_;
+}
+
+const char* ArrowStream::getError() const {
+  const char* lastError = arrowStream_->get_last_error(arrowStream_.get());
+  return lastError ? lastError : "";
```

Comment:
This method should only be called if there was an error, hence, we expect lastError to never be null. Can we check on that: VELOX_CHECK_NOT_NULL(lastError)? If not, let's return a message that clarifies that we couldn't get the error message instead of an empty string: `"<cannot get error message>"`.

Reply:
Hi Masha, the Arrow provided method (which is used in Gluten) returns nullptr for empty error msg. Is the second proposal better for this case?

Related link to Arrow code:
https://github.com/oap-project/arrow/blob/master/cpp/src/arrow/c/bridge.cc#L1663-L1666

Reply:
@rui-mo It is expected to receive nullptr from get_last_error if prior operation failed? If not, we can add a check that lastError is not null, otherwise, let's return `"<cannot get error message>"` or similar.

Reply:
Yes, you're right. Error msg only become empty when there is no exception, and in that case, getError method should not be called. Added a check to ensure lastError is not null.